### PR TITLE
OEL-786: Prepare spacing between paragraphs.

### PR DESCRIPTION
**Pull request by escuriola**
Jira issue(s):
- https://citnet.tech.ec.europa.eu/CITnet/jira/browse/OEL-786

### CODE REVIEW

**Review by drishu — CHANGES_REQUESTED** [review]
Nice, but we need to test this somehow, we have a paragraphs functional test with everything set up, we could profit and just add a test there that asserts the count of this class in the page

File: tests/src/Functional/ParagraphsTest.php [hunk 1]
```diff
@@ -48,6 +48,12 @@ class ParagraphsTest extends BrowserTestBase {
   protected function setUp(): void {
     parent::setUp();
 
+    \Drupal::service('theme_installer')->install(['oe_bootstrap_theme']);
```
Comment (drishu) [inline tests/src/Functional/ParagraphsTest.php]: remove this, rather add oe_bootstrap_theme to $defaultTheme

File: tests/src/Functional/ParagraphsTest.php [hunk 2]
```diff
@@ -48,6 +48,12 @@ class ParagraphsTest extends BrowserTestBase {
...
+    \Drupal::configFactory()
```
Comment (drishu) [inline tests/src/Functional/ParagraphsTest.php]: remove this

File: tests/src/Functional/ParagraphsTest.php [hunk 3]
```diff
@@ -151,6 +157,49 @@ public function testSocialMediaFollowParagraph(): void {
     $this->assertSession()->pageTextContains('More channels');
   }
 
+  /**
+   * Test paragraphs are wrapped with a spacing.
+   */
+  public function testSpacingBetweenParagraphs(): void {
```
Comment (drishu) [inline tests/src/Functional/ParagraphsTest.php]: rather then doing this we could just add the div count to testLinksBlockParagraph and testSocialMediaFollowParagraph

File: tests/src/Functional/ParagraphsTest.php [hunk 4]
```diff
@@ -48,6 +48,12 @@ class ParagraphsTest extends BrowserTestBase {
...
 
+    \Drupal::service('theme_installer')->install(['oe_bootstrap_theme']);
+    \Drupal::configFactory()
+      ->getEditable('system.theme')
+      ->set('default', 'oe_bootstrap_theme')
+      ->save();
+
     // Rebuild the ui_pattern definitions to collect the ones provided by
     // oe_bootstrap_theme itself.
     \Drupal::service('plugin.manager.ui_patterns')->clearCachedDefinitions();
```
Comment (drishu) [inline tests/src/Functional/ParagraphsTest.php]: try also without this line, dont think its needed

File: includes/menu.inc [hunk 5]
```diff
@@ -76,16 +76,9 @@ function oe_bootstrap_theme_preprocess_links__dropbutton(&$variables) {
...
   $variables['split'] = FALSE;
-  if (isset($link['link']) && ($url = $link['link']['#url'])) {
-    $button = $link['link'];
-
-    if ($variables['split'] = $url->getRouteName() !== '<nolink>') {
-      $button['#options']['attributes']['class'][] = 'btn';
-      $button['#options']['attributes']['class'][] = 'btn-sm';
-      $button['#options']['attributes']['class'][] = 'btn-outline-primary';
-    }
+  if (isset($link['text'])) {
```
Comment (drishu) [inline includes/menu.inc]: what are these changes dropbutton and dowpdown changes for ? seems out of scope of this ticket (if these are needed do report in a relevant ticket

File: modules/oe_bootstrap_theme_paragraphs/oe_bootstrap_theme_paragraphs.module [hunk 6]
```diff
@@ -24,6 +24,28 @@ use Drupal\oe_bootstrap_theme\ValueObject\ImageValueObject;
 use Drupal\oe_media_iframe\Plugin\media\Source\Iframe;
 use Drupal\paragraphs\Entity\Paragraph;
 
+/**
+ * Implements hook_theme_suggestions_HOOK_alter().
+ *
+ * Adds markup-free template suggestion to all paragraph fields.
```
Comment (drishu) [inline modules/oe_bootstrap_theme_paragraphs/oe_bootstrap_theme_paragraphs.module]: I think this comment here makes no sense

**Review by drishu — CHANGES_REQUESTED** [review]

File: includes/menu.inc (hunk 5 above)
Comment (escuriola) [reply includes/menu.inc]: This is needed for testing. With oe_bootstrap_theme enabled with this code when you are in a create node page  the list of available paragraphs is empty.

Comment (escuriola) [reply includes/menu.inc]: Created a new ticket: https://citnet.tech.ec.europa.eu/CITnet/jira/browse/OEL-961

**Review by drishu — DISMISSED** [review]

**Review by drishu — DISMISSED** [review]

File: templates/overrides/field/field--entity-reference-revisions--paragraph.html.twig [hunk 7]
```diff
@@ -0,0 +1,13 @@
+{#
+/**
+ * @file
+ * Theme override for the entity reference revisions field wich allowed to use paragraphs.
```
Comment (abel-santos-corral) [inline templates/overrides/field/field--entity-reference-revisions--paragraph.html.twig]:
```suggestion
 * Theme override for the entity reference revisions field which allowed to use paragraphs.
```

File: modules/oe_bootstrap_theme_paragraphs/oe_bootstrap_theme_paragraphs.module [hunk 8]
```diff
@@ -24,6 +24,23 @@ use Drupal\oe_bootstrap_theme\ValueObject\ImageValueObject;
...
+ */
+function oe_bootstrap_theme_paragraphs_theme_suggestions_field_alter(array &$suggestions, array $variables): void {
+  $element = $variables['element'];
+
+  // Do not output field labels and wrapping markup for paragraph fields.
+  if (isset($element['#entity_type']) && $element['#entity_type'] === 'paragraph') {
+    // Prepend the new suggestion to the list. This will put it right after the
+    // default field template. By doing this we allow to override single
+    // fields, while keeping all the rest markup-free.
+    array_unshift($suggestions, 'field__bare');
```
Comment (abel-santos-corral) [inline modules/oe_bootstrap_theme_paragraphs/oe_bootstrap_theme_paragraphs.module:41]: @drishu we brought this from OE Theme, we understand is necessary in OE Bootstrap theme. Right?

Another question is that it could be good to add an assertion in the test at EntityReferenceRevisionTest. We must ensure this suggestion is applied and no labels or attributes are added (because twig field__bare is used.). Is correct this assumption?

Comment (drishu) [reply modules/oe_bootstrap_theme_paragraphs/oe_bootstrap_theme_paragraphs.module:41]: yes, field__bare is good.
"ensure this suggestion is applied" since we check there for my-4 and find it, it means that the suggestion is applied, and that will be enough for now

File: tests/src/Kernel/Paragraphs/EntityReferenceRevisionTest.php [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+
+    $build = $node->get('test_paragraphs')->view();
+    $html = $this->render($build);
+    $crawler = new Crawler($html);
+
+    $wrappers = $crawler->filter('div.my-4');
+    $this->assertCount(2, $wrappers);
+    // Assertions for field bare twig.
+    $rich_text = $wrappers->eq(1);
+    $this->assertCount(1, $rich_text);
```
Comment (drishu) [inline tests/src/Kernel/Paragraphs/EntityReferenceRevisionTest.php]: if we assert above for count 2 its natural that there is a eq(1), this line is redundant

File: tests/src/Kernel/Paragraphs/EntityReferenceRevisionTest.php [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+    ]);
+    $node->save();
+
+    $build = $node->get('test_paragraphs')->view();
+    $html = $this->render($build);
+    $crawler = new Crawler($html);
+
+    $wrappers = $crawler->filter('div.my-4');
+    $this->assertCount(2, $wrappers);
+    // Assertions for field bare twig.
```
Comment (drishu) [inline tests/src/Kernel/Paragraphs/EntityReferenceRevisionTest.php]: checking that there is not title is fine, but this test is not about field bare, the implementation behind is not important (describing it is actually a mistake), its the resulted rendering that we check for, so please remove this comment.

File: tests/src/Kernel/Paragraphs/EntityReferenceRevisionTest.php [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+    $rich_text = $wrappers->eq(1);
+    $this->assertCount(1, $rich_text);
+    $title_rich_text = $rich_text->filter('div.oe-rich-text__field-oe-title');
+    $this->assertCount(0, $title_rich_text);
+    $text_rich_text = $rich_text->filter('div.oe-rich-text__field-oe-text-long');
+    $this->assertCount(0, $text_rich_text);
```
Comment (drishu) [inline tests/src/Kernel/Paragraphs/EntityReferenceRevisionTest.php]: since we don't sent the value of this field its natural that it doesn't appear no sense in checking that

File: tests/src/Kernel/Paragraphs/EntityReferenceRevisionTest.php [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+    // Assertions for field entity reference revisions twig.
```
Comment (drishu) [inline tests/src/Kernel/Paragraphs/EntityReferenceRevisionTest.php]: again, don't mention implementation, the behind the scenes can change, its the result that we need to check
```suggestion
    // Assert the paragraphs where rendered.
```

**Review by drishu — CHANGES_REQUESTED** [review]

**Review by drishu — APPROVED** [review]

**Review by abel-santos-corral — APPROVED** [review]
